### DISCUSSION TITLE
Can K3S single node setup be used for production grade system?

### DISCUSSION THREAD

**Discussion by ritesh1079** [k3s-io/k3s#2988] [Q&A]
Hi,

Can K3S single node setup(Server + Agent on same node) be used for production grade system?
 OR
is it only recommended for development setup?

Will it cause any performance or networking issues?

Min. system configuration, which I am having is:
RAM: 8GB
CPU: Dual Core, 2GHz

*****edited with more details on the scenario***
I am looking k3s deployment for edge device.

Currently I am evaluating running docker vs k3s in edge setup. Since k3s is coming lots of out of the box features like load balancing, ingress etc. I am more inclined towards k3s but wondering about it’s reliability, stability and performance in single node cluster. I did not find any recommendations on internet on this. Also, k3s documentation talks about at least two nodes setup. So want to double check on this before taking this decision.

Thanks!!

**Comment by DataHearth** [comment]
Hello there! I'm not an expert on that but a single node K3S is really light IMO for production environment.
I think you could use it for home production (like your slef hosted services,  website, etc), but for real production like enterprise or high usage services, where performance and stability is the key, you should really go for HA cluster :D.

So, yeah it is recommended for development ^^.

When it comes to performances issues, depending on what your running on the cluster, you might experience some troubleshoots (never done something similar tho).

Cheers !

**Reply by brandond** [reply]
Production-grade doesn't necessarily mean HA; making that decision would be up to the user based on the requirements of your specific use case. For edge or other resource-constrained deployments, you may not have more than a single node available in a particular location.

K3s doesn't make any decisions that would make it less production-ready than a single-node kubeadm-based Kubernetes cluster. It is intended to be **more** suitable for single-node deployments, since we we optimize for smaller nodes by using less CPU/Memory/Disk resources... but you can still scale it up to large multi-node clusters, just as you can with upstream Kubernetes.

**Reply by ritesh1079** [reply]
Thanks for the response!!
Exactly, I am looking k3s deployment for edge device.
Currently I am evaluating running docker vs k3s in edge setup. Since k3s is coming lots of out of the box features like load balancing, ingress etc. I am more inclined towards k3s but wondering about it’s reliability, stability and performance in single node cluster. I did not find any recommendations on internet on this. Also, k3s documentation talks about at least two nodes setup. So want to double check on this before taking this decision.

**Reply by i5Js** [reply]
I recommend at least one master and one worker. I've an HA cluster, two master and 5 workers, all raspberry and I can't be more happy :)

**Reply by DataHearth** [reply]
@ritesh1079 Well as @brandond  said, k3s isn't really less productive. I would also say that it is not less efficient and stable. k3s is for me, a great choice when you have a single node cluster with less requirements and/or less hardware power. 
I would also that it is more stable for me in terms of management for small production environment compare to the same instance of a k8s cluster because k8s is designed to be more efficient and stable in much bigger environment IMO.

To answer your question, here's my advice, stay with k3s :D. It is the best choice for me when I look at what you said ;).

**Reply by ritesh1079** [reply]
@DataHearth, @brandond, @i5Js, Thanks for your inputs!!
